### ISSUE TITLE
Razer DA Elite - Save DPI

### ISSUE THREAD

**Issue by furai** [polychromatic/polychromatic#203]
Hello,

Is it possible to save mouse DPI for the Razer DA Elite? It seems that it resets for me every reboot. Is this connected with the issue mentioned in Readme that mouse and mouspads don't have profiles?

Cheers,
Furai

**Comment by lah7** [comment]
At the moment, Polychromatic relies on the hardware to retain its DPI state between reboots. Mice like the Mamba do, but others like DeathAdder Elite "reset" and solely rely on software (specifically, Razer Synapse) to set the DPI.

In future, yes. With new profile support, it will be easy to create a preset and apply those settings on start-up.

For now (v0.3.12), it is possible to create a mini Python script to interface with the OpenRazer library directly - create this file and add it to your Startup Applications:
```
#!/usr/bin/python3
import openrazer.client
a = openrazer.client.DeviceManager()
b = a.devices[0] // Assumes DA Elite is the only and first device
b.dpi = (800,800)
```

**Comment by furai** [comment]
Oh, neat. So TBH I don't really need the whole polychromatic? I don't care for all the colours and stuff like that. I just want to set some stuff for the mouse and forget about it.

Can I just remove polychromatic altogether and manage openrazer through a script? Are there any docs on that?

**Comment by lah7** [comment]
Yes, if you prefer. There are no "finished" docs for OpenRazer's Python library right now, but you can poke around in a `python3` interpreter like so. 

```
~$ python3
Python 3.5.2 (default, Nov 12 2018, 13:43:14) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import openrazer.client
>>> devices = openrazer.client.DeviceManager().devices
>>> mouse = devices[0]
>>> mouse.name
'Razer Mamba Tournament Edition'
>>> mouse.dpi
(800, 800)
>>> mouse.fx.
```

Pressing <kbd>Tab</kbd> after a dot will list the possible functions. Ignore underscores.
```
mouse.fx.__class__(             mouse.fx.__subclasshook__(
mouse.fx.__delattr__(           mouse.fx.__weakref__
mouse.fx.__dict__               mouse.fx._capabilities
mouse.fx.__dir__(               mouse.fx._custom_lighting_dbus
mouse.fx.__doc__                mouse.fx._dbus
mouse.fx.__eq__(                mouse.fx._lighting_dbus
mouse.fx.__format__(            mouse.fx.advanced
mouse.fx.__ge__(                mouse.fx.breath_dual(
mouse.fx.__getattribute__(      mouse.fx.breath_random(
mouse.fx.__gt__(                mouse.fx.breath_single(
mouse.fx.__hash__(              mouse.fx.breath_triple(
mouse.fx.__init__(              mouse.fx.has(
mouse.fx.__le__(                mouse.fx.misc
mouse.fx.__lt__(                mouse.fx.none(
mouse.fx.__module__             mouse.fx.reactive(
mouse.fx.__ne__(                mouse.fx.ripple(
mouse.fx.__new__(               mouse.fx.ripple_random(
mouse.fx.__reduce__(            mouse.fx.spectrum(
mouse.fx.__reduce_ex__(         mouse.fx.starlight_dual(
mouse.fx.__repr__(              mouse.fx.starlight_random(
mouse.fx.__setattr__(           mouse.fx.starlight_single(
mouse.fx.__sizeof__(            mouse.fx.static(
mouse.fx.__str__(               mouse.fx.wave(
```

For example:
```
>>> mouse.fx.static(0, 255, 0)     # Sets effect to static green (RGB 0,255,0)
```

**Comment by furai** [comment]
Thanks a ton! I'll have a look into it. :)

**Comment by furai** [comment]
BTW, would there be a way to change DA Elite built in DPI levels you change with a button press? Or rebind this keys to something different?

**Comment by lah7** [comment]
Unfortunately not as neither OpenRazer or Polychromatic support key rebinding right now.

**Comment by furai** [comment]
What about changing DPI levels?

**Comment by lah7** [comment]
It's the software (Razer Synapse) that takes control of the DPI buttons and sets the DPI according to [the values set for each stage in the software](https://www.overclock.net/photopost/data/1554986/1/1d/1df48b99_DeathadderElite_DefaultDPI-Steps.jpeg). I don't know if the firmware itself stores each DPI level, but I would **guess** it's only knows the last DPI value and uses the factory defaults for each stage when the "Synapse" driver isn't present.

**Comment by furai** [comment]
Thanks yet again for your time and invaluable information that you have provided me with.

**Comment by redlet** [comment]
@lah7 
`b.dpi = (800,800)`
why two values?

**Comment by lah7** [comment]
> Question 2: is it possible to have a script run if the DPI accidentally changes?

Yes. You could create a script to monitor the DPI every minute and set the DPI value again if it changes. Though, this would be inefficient as this is **polling** and not **event** driven, so it would waste CPU cycles.

To my knowledge, I don't know how to listen to the mouse's DPI buttons (event-driven).

**Comment by lah7** [comment]
**Stale issue,** let's get it up to speed:
* OpenRazer daemon has a "persistence" feature. 
    * Writing a custom start-up script isn't required, but it does provide full flexibility.
* OpenRazer daemon doesn't support key remapping.
    * Use a generic remapper utility like [Input Remapper](https://github.com/sezanzeb/input-remapper) instead.
* Known issue that the DPI button click might not be emitted, see here:
   *  https://github.com/openrazer/openrazer/issues/201
* Polychromatic has support for saving DPI stages to the hardware, if the mouse (+ OpenRazer) supports it.
* Some newer mice have a "profile" button underneath (with a colour LED), which isn't supported in OpenRazer. I believe that's [onboard memory section.](https://github.com/openrazer/openrazer/wiki/Unknown-commands#on-board-memory) Only the default (white?) profile is supported.

> In future, yes. With new profile support, it will be easy to create a preset and apply those settings on start-up.

That's still to come. I found the term "profile" to became ambiguous. "Presets" is still a feature I'd like to have, which is just a fancier way to specify settings under different scenarios.

I'll close the issue since the initial DPI questions have been resolved.
